TSM: Rename while file is in Use in compact_test

I am trying to implement mmap and tsm compatibility to Windows via PR #4981 to fix #4358. 
But MustWriteTSM in compact_test.go is trying to rename a file which is in use (a possible file leak scenario, not supported by Windows) is preventing me from proceeding. Could somebody here help me to resolve the issue?

    >go test -v ./tsdb/engine/tsm1 > %temp%\gotest.txt 2>&1

         --- FAIL: TestCompactor_Compact (0.00s)
     panic: create tsm file: rename C:\Users\Ady\AppData\Local\Temp\tsm1-test087429749\tsm1test014002000 C:\Users\Ady\AppData\Local\Temp\tsm1-test087429749\000000001-000000001.tsm: The process cannot access the file because it is being used by another process. [recovered]
     	panic: create tsm file: rename C:\Users\Ady\AppData\Local\Temp\tsm1-test087429749\tsm1test014002000 C:\Users\Ady\AppData\Local\Temp\tsm1-test087429749\000000001-000000001.tsm: The process cannot access the file because it is being used by another process.

[gotest.txt](https://github.com/influxdb/influxdb/files/59701/gotest.txt)

The root cause is in below code `compact_test.go:761` where file handle returned by `MustTempFile` is not closed before rename operation, which is failing in Windows. Simply closing that handle before rename fails the next write operation.

     func MustWriteTSM(dir string, gen int, values map[string][]tsm1.Value) string {
    	f := MustTempFile(dir)
    	newName := filepath.Join(filepath.Dir(f.Name()), tsmFileName(gen))
    	if err := os.Rename(f.Name(), newName); err != nil {
    		panic(fmt.Sprintf("create tsm file: %v", err))
    	}

--- a/tsdb/engine/tsm1/compact_test.go
+++ b/tsdb/engine/tsm1/compact_test.go
@@ -981,9 +981,23 @@ func MustWALSegment(dir string, entries []tsm1.WALEntry) *tsm1.WALSegmentReader 
 
 func MustWriteTSM(dir string, gen int, values map[string][]tsm1.Value) string {
 	f := MustTempFile(dir)
-	newName := filepath.Join(filepath.Dir(f.Name()), tsmFileName(gen))
-	if err := os.Rename(f.Name(), newName); err != nil {
+	oldName := f.Name()
+
+	// Windows can't rename a file while it's open.  Close first, rename and
+	// then re-open
+	if err := f.Close(); err != nil {
+		panic(fmt.Sprintf("close temp file: %v", err))
+	}
+
+	newName := filepath.Join(filepath.Dir(oldName), tsmFileName(gen))
+	if err := os.Rename(oldName, newName); err != nil {
 		panic(fmt.Sprintf("create tsm file: %v", err))
+	}
+
+	var err error
+	f, err = os.OpenFile(newName, os.O_RDWR, 0666)
+	if err != nil {
+		panic(fmt.Sprintf("open tsm files: %v", err))
 	}
 
 	w, err := tsm1.NewTSMWriter(f)

